pkg/compactor/compactor.go

```diff
@@ -0,0 +1,281 @@
+}
+
+func (c *Compactor) run() {
+	defer c.runner.Done()
+
+	// Run an initial compaction before starting the interval.
+	c.compactUsersWithRetries(c.ctx)
+
+	ticker := time.NewTicker(c.compactorCfg.CompactionInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.compactUsersWithRetries(c.ctx)
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Compactor) compactUsersWithRetries(ctx context.Context) {
+	retries := util.NewBackoff(ctx, util.BackoffConfig{
+		MinBackoff: c.compactorCfg.retryMinBackoff,
+		MaxBackoff: c.compactorCfg.retryMaxBackoff,
+		MaxRetries: c.compactorCfg.CompactionRetries,
+	})
+
+	c.compactionRunsStarted.Inc()
+
+	for retries.Ongoing() {
+		if err := c.compactUsers(ctx); err == nil {
+			c.compactionRunsCompleted.Inc()
+			return
+		}
+
+		retries.Wait()
+	}
+
+	c.compactionRunsFailed.Inc()
+}
+
+func (c *Compactor) compactUsers(ctx context.Context) error {
+	level.Info(c.logger).Log("msg", "discovering users from bucket")
+	users, err := c.discoverUsers(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to discover users from bucket", "err", err)
+		return err
+	}
+	level.Info(c.logger).Log("msg", "discovered users from bucket", "users", len(users))
+
+	for _, userID := range users {
+		// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		level.Info(c.logger).Log("msg", "starting compaction of user blocks", "user", userID)
+
+		if err = c.compactUser(ctx, userID); err != nil {
+			level.Error(c.logger).Log("msg", "failed to compact user blocks", "user", userID, "err", err)
+			continue
+		}
+
+		level.Info(c.logger).Log("msg", "successfully compacted user blocks", "user", userID)
+	}
+
+	return nil
+}
+
+func (c *Compactor) compactUser(ctx context.Context, userID string) error {
+	bucket := cortex_tsdb.NewUserBucketClient(userID, c.bucketClient)
+
+	syncer, err := compact.NewSyncer(
+		c.logger,
+		nil, // TODO(pracucci) we should pass the prometheus registerer, but we would need to inject the user label to each metric, otherwise we have clashing metrics
+		bucket,
+		c.compactorCfg.ConsistencyDelay,
+		c.compactorCfg.BlockSyncConcurrency,
+		false, // Do not accept malformed indexes
+		true,  // Enable vertical compaction
+		[]*relabel.Config{})
+	if err != nil {
+		return errors.Wrap(err, "failed to create syncer")
+	}
+
+	compactor, err := compact.NewBucketCompactor(
+		c.logger,
+		syncer,
+		c.tsdbCompactor,
+		path.Join(c.compactorCfg.DataDir, "compact"),
+		bucket,
+		// No compaction concurrency. Due to how Cortex works we don't
+		// expect to have multiple block groups per tenant, so setting
+		// a value higher than 1 would be useless.
+		1,
+	)
+	if err != nil {
+		return errors.Wrap(err, "failed to create bucket compactor")
+	}
+
+	return compactor.Compact(ctx)
+}
+
+func (c *Compactor) discoverUsers(ctx context.Context) ([]string, error) {
+	users := make([]string, 0)
```

Comment:
Nit. Just `var users []string` would suffice. I am not sure how Go works in this case, but I vaguely remember that there are no allocations if just declared and not appended to it yet (also, it looks cleaner).

Reply:
Changed even if such kind of changes for any non-hot path are not really a valuable thing.